types/stream-to-promise/index.d.ts

```diff
@@ -4,13 +4,14 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 /// <reference types="node"/>
+import * as Stream from 'stream';
 
 /**
  * Return a promise which resolves when the input stream ends
  * @param stream The input stream
  * @returns A promise containing the stream output as a buffer
  */
-declare function streamToPromise(stream: NodeJS.ReadableStream): Promise<Buffer>;
+declare function streamToPromise(stream: Stream.Duplex | NodeJS.ReadableStream): Promise<Buffer>;
```

Comment:
Based on the code you linked, this would be cleaner with `stream: NodeJS.ReadableStream | NodeJS.WritableStream`

Reply:
Thanks for the response!

I'm not sure that would work because that would conflict with the next overload in the file which is: `stream: NodeJS.WritableStream`.

Actually now that I look at this more closely the issue isn't as simple as I thought. I was under the impression that TypeScript resolved overloads using a simple top-down approach, returning the first overload to match. But this example doesn't seem to follow those rules:

```ts
import * as Stream from 'stream';
function streamToPromise(stream: NodeJS.ReadableStream): Promise<Buffer>;
//function streamToPromise(stream: NodeJS.WritableStream): Promise<void>;
function streamToPromise(stream: any): any {}

// This will raise an error if the second overload is uncommented.
const fromDuplex: Promise<Buffer> = streamToPromise(0 as any as Stream.Duplex);
```

We can be sure that `Stream.Duplex` is structurally compatible with `NodeJS.ReadableStream` because this example compiles fine when the second overload is commented out. If you uncomment the `WritableStream` overload it fails to compile because it selects the `Promise<void>` overload which can't be assigned to `Promise<Buffer>`. I thought this would be impossible because it's declared after the correct overload.

So from here we could either work around the issue with this apparently redundant declaration:
```ts
export default function streamToPromise(stream: NodeJS.ReadableStream | Stream.Readable): Promise<Buffer>;
export default function streamToPromise(stream: NodeJS.WritableStream | Stream.Writable): Promise<void>;
```

..or I could open an issue on the TypeScript github before putting workarounds in place for issues I don't understand.

Reply:
Ah, my mistake - I didn't see the context of the change, so I didn't notice the next line had that overload.
What is the issue you're trying to fix here? Just that you want to be able to call this function with a `Stream.Duplex`?

Reply:
Sorry I could have explained this better in the original comment. This simple module returns a promise that is resolved when a given stream finishes. In the case of a readable stream it will resolve to the contents that were read from the stream. In the case of a writable stream it just returns a void promise. In the case of a duplex stream the readable behavior is prioritized.

The problem is that for some reason TypeScript is choosing the second overload (writable -> Promise<void>) over the first overload (readable -> Promise<Buffer>). This seems to me like a bug in TypeScript. I've pushed a workaround which explicitly declares the overload as `NodeJS.ReadableStream | Stream.Readable`, but this shouldn't be necessary since the two types are structurally compatible.